subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_poll.c

```diff
@@ -0,0 +1,473 @@
+
+#if defined(CONFIG_NRF_CLOUD_REST)
+	err = nrf_cloud_rest_disconnect(ctx->rest_ctx);
+#elif defined(CONFIG_NRF_CLOUD_COAP)
+	err = nrf_cloud_coap_disconnect();
+#else
+	err = -ENOTSUP;
+#endif
+#if defined(CONFIG_LTE_LINK_CONTROL)
+	(void)lte_lc_power_off();
+#endif
+
+	return err;
+}
+
+static void cleanup(void)
+{
+#if defined(CONFIG_NRF_CLOUD_REST)
+	nrf_cloud_rest_fota_job_free(&job);
+#elif defined(CONFIG_NRF_CLOUD_COAP)
+	nrf_cloud_coap_fota_job_free(&job);
+#endif
+}
+
+/****************************************************/
+/* End of transport-specific wrappers.              */
+/****************************************************/
+static void http_fota_dl_handler(const struct fota_download_evt *evt)
+{
+	LOG_DBG("evt: %d", evt->id);
+
+	switch (evt->id) {
+	case FOTA_DOWNLOAD_EVT_FINISHED:
+		LOG_INF("FOTA download finished");
+		fota_status = NRF_CLOUD_FOTA_SUCCEEDED;
+		k_sem_give(&fota_download_sem);
+		break;
+	case FOTA_DOWNLOAD_EVT_ERASE_PENDING:
+	case FOTA_DOWNLOAD_EVT_ERASE_TIMEOUT:
+		LOG_INF("FOTA download erase ongoing");
+		break;
+	case FOTA_DOWNLOAD_EVT_ERASE_DONE:
+		LOG_DBG("FOTA download erase done");
+		break;
+	case FOTA_DOWNLOAD_EVT_ERROR:
+		LOG_INF("FOTA download error: %d", evt->cause);
+
+		fota_status = NRF_CLOUD_FOTA_FAILED;
+		fota_status_details = FOTA_STATUS_DETAILS_DL_ERR;
+
+		if (evt->cause == FOTA_DOWNLOAD_ERROR_CAUSE_INVALID_UPDATE) {
+			fota_status = NRF_CLOUD_FOTA_REJECTED;
+			if (nrf_cloud_fota_is_type_modem(job.type)) {
+				fota_status_details = FOTA_STATUS_DETAILS_MDM_REJ;
+			} else {
+				fota_status_details = FOTA_STATUS_DETAILS_MCU_REJ;
+			}
+		} else if (evt->cause == FOTA_DOWNLOAD_ERROR_CAUSE_TYPE_MISMATCH) {
+			fota_status_details = FOTA_STATUS_DETAILS_MISMATCH;
+		}
+		k_sem_give(&fota_download_sem);
+		break;
+	case FOTA_DOWNLOAD_EVT_PROGRESS:
+		LOG_DBG("FOTA download percent: %d%%", evt->progress);
+		break;
+	default:
+		break;
+	}
+}
+
+static bool pending_fota_job_exists(void)
+{
+	return (pending_job.validate != NRF_CLOUD_FOTA_VALIDATE_NONE);
+}
+
+static void process_pending_job(struct nrf_cloud_fota_poll_ctx *ctx)
+{
+	bool reboot_required = false;
+	int ret;
+
+	ret = nrf_cloud_pending_fota_job_process(&pending_job, &reboot_required);
+
+	if ((ret == 0) && reboot_required) {
+		/* Save validate status and reboot */
+		(void)nrf_cloud_fota_settings_save();
+
+		if (ctx->reboot_fn) {
+			ctx->reboot_fn(FOTA_REBOOT_REQUIRED);
+		}
+	}
+}
+
+int nrf_cloud_fota_poll_init(struct nrf_cloud_fota_poll_ctx *ctx)
+{
+	int err;
+
+	if (!check_ctx(ctx)) {
+		LOG_ERR("Invalid ctx");
+		return -EINVAL;
+	}
+
+	ctx->full_modem_fota_supported = false;
+
+	err = nrf_cloud_fota_settings_load(&pending_job);
+	if (err) {
+		return err;
+	}
+
+#if defined(CONFIG_NRF_CLOUD_FOTA_FULL_MODEM_UPDATE)
+	struct dfu_target_fmfu_fdev fmfu_dev_inf = {
+		.size = 0,
+		.offset = 0,
+		.dev = IS_ENABLED(CONFIG_DFU_TARGET_FULL_MODEM_USE_EXT_PARTITION) ?
+		       NULL : DEVICE_DT_GET_ONE(jedec_spi_nor)
```

Comment:
since this is in the common library now, the flash device name should be Kconfigurable?

Reply:
It looks like ATv2 does not make it Kconfigurable -- they assume `jedec_spi_nor`.

Reply:
Atv2 is for Nordic hardware.
This library should be usable by custom devices?

Reply:
They intend people to use Atv2 as the basis for their products, not just for Nordic hardware.

Reply:
I think it would be cleaner to stop supporting use of the external flash directly and just require using a partition, so the flash device can be managed elsewhere.

On the other hand, `subsys/net/lib/fota_download/src/util/fota_download_full_modem.c` does this:
`static const struct device *flash_dev = DEVICE_DT_GET_ANY(jedec_spi_nor);`
so maybe we should use the same if ext partitions are not used.

Reply:
atv2 is an application and nrf_cloud is library.
a user shouldn't need to modify the library to get it to work with their product.

Reply:
fota_download is a library as well

Reply:
this PR is for nrf_cloud library.

if you don't want to make it configurable just say no and add a note in the docs.

Reply:
that was easy 😅